### ISSUE TITLE
Defining "Experimental"/"Feature Flags"/?

### ISSUE THREAD

**Issue by CEHENKLE** [opensearch-project/.github#108]
### Is your feature request related to a problem?
We had a couple of features go out under "experimental" or "feature" flags in the last 2 releases.    But we haven't (among other things) defined:

1. What features can be released as experimental?  
2. What criteria do you need to meet to be able to release as experimental?
3. Before a feature is released experimentally, what is the plan to get off of "experimental"?
4. What is a feature flag and why do we use them?
5. What criteria do you need to meet to be able to turn a feature flag off?
6. What is the correct way to add these (YAML file?  Stability flag)?  Or how should people find them?
7. How can we scale testing when we have several experimental features in a release?

### What solution would you like?
An update on contributing to answer these questions (and any others that come up)

**Comment by seanneumann** [comment]
To expanding on testing, I'd like to add that "experimental" does not mean low quality. One of the primary purposes of experimental is to get feedback, not to get features out early. Identifying a clear quality bar would be great. For example, glaring bugs, incomplete copy or UI, lack of critical tests, and obvious performance issues.

**Comment by andrross** [comment]
> What is a feature flag and why do we use them?

Speaking for the opensearch-project/OpenSearch repo, I believe feature flags are intended to replace long-running feature branches. This allows for early integration of incomplete features into the main development branch and avoids major pain with merging a long-running feature branch. I believe we have overloaded this mechanism to use it for experimental features. As @seanneumann called out, "experimental" features shouldn't be incomplete or have obvious performance issues, but features under development are somewhat by definition incomplete. The bar for adding a new feature flag should be pretty simple, along the lines of 1) must not adversely impact any existing functionality and 2) must have a plan to complete the feature (proofs-of-concept and wild new ideas can be shared on separate branches and/or user forks). We definitely need to clearly define and enforce a policy here though.

For the broader concept of an "experimental" feature, one that is officially documented and announced to users, I'm coming around to the opinion that we need a new mechanism. I think such features are really in a different phase of their lifecycle than the feature flag. They are not incomplete but instead we're looking for early feedback before committing to long-term support promised by our semver policy.

tl;dr OpenSearch feature flags were intended to replace feature branches. A new mechanism is needed for releasing features for early feedback.

**Comment by dblock** [comment]
I like to think of "experimental" features as "working, but we're not sure we will want to keep it", rather than "incomplete, a lot of work remains to be done". That said, whichever way I personally interpret it, I think we can all agree that the bar for experimental features is lower in terms of various aspects of completeness: 1) the feature is not turned on by default, 2) APIs may not be following semver yet, 3) documentation may not be fully written and proof-read, 4) performance may not be at par with initial expectations, 5) performance regressions aren't yet treated as show-stopping bugs. 

Another way of thinking about experimental features is that they have not yet [crossed the chasm](https://en.wikipedia.org/wiki/Crossing_the_Chasm). They are aimed at early adopters that are willing to put up with 1), 2), 3), 4) and 5) above. Therefore, I think we definitely want feedback at all stages of all features, but in the experimental phase we'd want it from those who have the appetite to try them.

One place where I wouldn't want to lower expectations for experimental features is security. 

Question: where do we want this information? In https://github.com/opensearch-project/.github/blob/main/RELEASING.md or on openseach.org (https://github.com/opensearch-project/project-website)?

**Comment by wbeckler** [comment]
In terms of where to put this, when I have a hard time finding something, I put it back in the first place I checked. In the case of OpenSearch, maybe that place would be the cluster settings API, given that experimental features are cluster-wide. 

If the concept of "experimental" is to prevent overconfident dependency on the feature, maybe we should consider covering such features with "wet paint" signs. One way to do that would be to include "experimental" in the name the feature. For example, the [experimental remote store API](https://github.com/opensearch-project/OpenSearch/pull/3576) could have been named the "experimental remote store API" and the API endpoint would be 
`POST /_experimentalremotestore/_restore`
instead of 
`POST /_remotestore/_restore`

The related settings that go with this feature would also be prepended with "experimental." 

This would prevent a reliance on an experimental feature that could potentially be yanked.

**Comment by wbeckler** [comment]
With both experimental features and with feature flags, it's useful to set a hard limit on the number of such things floating around, as they carry a cost. Having a hard limit means that if someone wants to add one more, they need to go about removing one of the ones that is already hanging around, which such things inevitably do. The hard limit addresses the issue of someone adding something temporary and then disappearing and nobody else having the incentive to remove it. The person who wants to add a new one is the person who has the incentive to kill off an old one.

**Comment by dblock** [comment]
I think @wbeckler is onto something and experimental features should be product-specific, appear in https://github.com/opensearch-project/OpenSearch/blob/main/DEVELOPER_GUIDE.md, and user-facing documentation. We can add some high level approach for plugins to https://github.com/opensearch-project/opensearch-plugins/blob/main/CONVENTIONS.md, etc. 

I like the ideas such as `/_experimental/` for RESTful APIs, but I also don't feel that it's necessary to break users on purpose in upgrades.

**Comment by dblock** [comment]
In some offline conversation we talked about the fact that experimental features should be 1) disabled by default 2) a user should not be able to enable them via some API, ie. you must be able to reconfigure the cluster (e.g. via editing .yml file) to enable the experimental feature.

**Comment by macohen** [comment]
I agree that experimental features need to be disabled by default, cannot be enabled via API. I also think that security should be an issue all along, but is there a way to complete penetration tests reliably if the API is changing? 

My understanding is that a good pentest needs a stable API so if the API could change I think we need to consider that some of the security concerns may need to happen after a feature is experimental, but before the feature flag is enabled by default.

**Comment by dblock** [comment]
> I agree that experimental features need to be disabled by default, cannot be enabled via API. I also think that security should be an issue all along, but is there a way to complete penetration tests reliably if the API is changing?
> 
> My understanding is that a good pentest needs a stable API so if the API could change, then some of the security concerns may need to be considered after a feature is experimental, but before the feature flag is enabled by default.

Most experimental features don't seem to be changing APIs that much, but if they do, additional security reviews may be required. We can simply acknowledge that.
